Plane: scale angle gains in POSITION1 state with airspeed

tridge opened the pull request:
During POSITION1 we still have significant airspeed and using the VTOL Q_A_ANG_P gain may be too high when the fixed wing controller is still dominant and the fixed wing time constant equates to a much smaller angle P gain.
When tuning the VTOL controller we tune in hover, with low airspeed and correspondingly low fixed wing control. It is really not surprising that when we use those gains at high airspeed in POSITION1 that we sometimes see P gain oscillation

This is a simple change to keep using the fixed wing angle gains until we get to the POSITION2 state

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -843,7 +843,7 @@ void QuadPlane::run_esc_calibration(void)
 void QuadPlane::multicopter_attitude_rate_update(float yaw_rate_cds)
 {
-    bool use_multicopter_control = in_vtol_mode() && !tailsitter.in_vtol_transition();
+    bool use_multicopter_control = !use_fw_attitude_controllers();
```
This is not the same, we loose the `!tailsitter.in_vtol_transition()`

tridge replied:
fixed

IamPete1 replied:
I actually don't think we need this change at all. `multicopter_attitude_rate_update` is not called in pos1. Not that it might not be a valid change, but removing it and the tailsitter change it needed in `use_fw_attitude_controllers` makes this a 1 line change.

IamPete1 commented:
This is one of the fundamental problems with using two controllers to try and achieve the same target.

In the majority of cases the equivalent FW angle gains will be too low and the vehicle will fly very badly. 

Lower rate limits in both FW and VTOL should help, once the response is rate limited you can turn up the angle gain by a order of magnitude and it makes no difference. 

We could also enable the attitude control time constant to allow better input shaping, again this should calm everything down. 

Rather than turning on all of the clever input shaping in attitude control my preference would be to implement such input shaping in the FW control so we can inherit it keeping the target rates to the two controllers the same. If we just turn on the copter input shaping it will reduce the target copter is aiming for while FW is still trying to get to the original value.

tridge commented:
> This is one of the fundamental problems with using two controllers to try and achieve the same target.

what I'd really like to do is gain schedule the angle gain between the fixed wing gain and the VTOL gain based on the airspeed, but I don't see a way to do that with the existing AC_AttitudeControl structure

> In the majority of cases the equivalent FW angle gains will be too low and the vehicle will fly very badly.

I don't think it does - low angle gains slows things down, which is only a problem if the next loop out is demanding rapid changes. So it would be more of a problem if we were running the full pos controller, but with the current structure I don't see it being an issue, and SITL testing in RF9 doesn't yet show any issues

IamPete1 commented:
I have had a little play and think, hopefully this explains what this PR does and the behavior change. 

This plots the desired rates of the FW controller in red and VTOL in blue:

![image](https://user-images.githubusercontent.com/33176108/194731656-fc612f93-ad66-40f0-9433-60d22aa702b2.png)

1. FW flight, FW controller running as usual, VTOL controller not running
2. VTOL air-brake. VTOL controllers slaved to FW at rate level.
3. POS1, This is the changed section. On master this is FW slaved to VTOL at rate level. Now we just run both controllers, no slaving but both trying for the same angle.
4. POS2 FW slaved to VTOL.

Section 3 also shows the problem this is trying to overcome. The VTOL target rates are of higher magnitude than those of the FW controller, so feeding those targets in to the FW could result in oscillation.

IamPete1 commented:
I think the combined approach we were discussing can be summed up like this:

Currently we have two angle controllers and two rate controllers, one for FW and one for VTOL. We pick which to run depending on what were doing. 

The combined approach would be to have a single angle controller, with a variable gain depending on what were doing. This would then feed into the two rate controllers.

tridge commented:
@IamPete1 I discussed with @lthall and he is happy with the addition of an AC_AttitudeControl API that takes a vector3f of scalings for the angle P gains for roll/pitch/yaw, applies for one loop, auto-resets to 1.0 after the loop

tridge commented:
@IamPete1 I think we should include this simpler patch in 4.3.1 however, if testing goes well

IamPete1 commented:
I have not yet tested this, but I think `Q_TRIM_PITCH` will cause issues here. FW is aiming for angle X in normal AHRS view and VTOL is aiming for X in VTOL view. Those two angles may not be the same due to `Q_TRIM_PITCH` so it will cause the controllers to fight. 

Unfortunately I don't think there is any great way to fix this without turning into quite a big patch that we may not want to back-port.

tridge commented:
> Those two angles may not be the same due to the trims so it will cause the controllers to fight.

there is only one angle controller at a time, the other is a rate controller only

IamPete1 commented:
This runs two angle controllers in POS1, that is why we see different target rates in PIDP and PIQP.

tridge commented:
> This runs two angle controllers in POS1, that is why we see different target rates in PIDP and PIQP.

hmm, you're right, and that was not the intention. I've changed it so it does properly slave the VTOL rates to fw rates

tridge commented:
with the change:
![image](https://user-images.githubusercontent.com/831867/194779407-694f7a97-043d-4409-9462-cea0e57caceb.png)

tridge commented:
I suspect we're adding one loops worth of lag with this change

IamPete1 commented:
This also breaks the overshoot stuff we added recently, since the FW controller will be trying to reverse while the yaw does odd stuff. 

I wonder if we might be better off doing the full gain scaling way and forgetting about keeping it small for a backport. I think the vehicles this has been seen on should fly OK with VTOL lower angle gains until we can get the full solution done. 

Its very easy to accidentally break stuff in exciting ways in this bit of the code.

tridge commented:
@IamPete1 certainly worth trying an implementation of the scaling and see how it goes

tridge commented:
@IamPete1 I've redone this PR with the scaling approach, tested in SITL

IamPete1 commented on libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:
```diff
@@ -988,25 +990,32 @@ Vector3f AC_AttitudeControl::update_ang_vel_target_from_att_error(const Vector3f
     }
+
+    // reset angle P scaling
+    _angle_P_scale = Vector3f{1,1,1};
```
This reset might be better in `rate_controller_run`, I would be us-surprised to find we come though here twice on plane sometimes.

bnsgeyer replied:
IMO it should be fine resetting the scalar here.  The rate correction for the angle error is added to the feedforward rate before being sent to the rate controller.  So to my knowledge, the rate correction is not calculated again.

lthall replied:
I doubt plane is using this twice but if it is that needs to be fixed. I think it is safe to assume that any users of the Attitude controllers aren't doing anything silly.

tridge replied:
discussed with leonard and ended up with both a set and a set mult call

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -4309,4 +4316,45 @@ bool QuadPlane::landing_with_fixed_wing_spiral_approach(void) const
+    const float angP_scale = linear_interpolate(mc_angP, fw_angP,
+                                                aspeed,
+                                                low_airspeed, plane.aparm.airspeed_min) / mc_angP;
+    const Vector3f gain_scale{angR_scale, angP_scale, 1.0};
```
Why not do yaw?

tridge replied:
no obvious yaw time constant to use, plus yaw hasn't been an issue in logs

IamPete1 reviewed (commented):
I think we should probably also scale max rates.

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -4309,4 +4316,45 @@ bool QuadPlane::landing_with_fixed_wing_spiral_approach(void) const
+        return;
+    }
+
+    const float angR_scale = linear_interpolate(mc_angR, fw_angR,
```
I wonder if it should be airspeed squared. Linear is probably close enough tho.

tridge replied:
linear seems to most obvious and simplest, its hard to reason about the combination of the two controllers

tridge commented:
> I think we should probably also scale max rates.

the max rates tend to be similar, so I don't think it is worth scaling them

IamPete1 commented:
I added some logging of the gains to really see what is going on.

![image](https://user-images.githubusercontent.com/33176108/196005655-9123908f-821a-4905-9791-0e749276c48e.png)

Two points of interest:

- Change in gain is quite "steppy" because of the difference in update rate between the attitude loop and the airspeed loop.
- Because we change into POS2 before reaching `low_airspeed` there is a step change from 0.87 to 1 when we stop scaling. Due to the target filters this seems to not make it out to the motors, so maybe its not that important.

I also had a go at back-calculating this value from the roll/pitch error. It doesn't really work, I think if we might be interested we need to log either it directly, or the angle errors. (Unless I have my calculation wrong.)

![image](https://user-images.githubusercontent.com/33176108/196006309-134ed3f9-af98-4627-b1c0-46c792fe78ec.png)

![image](https://user-images.githubusercontent.com/33176108/196006316-4cb6d0f1-c743-4397-aafb-65c6879fd4f0.png)

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -4309,4 +4316,45 @@ bool QuadPlane::landing_with_fixed_wing_spiral_approach(void) const
+        return;
+    }
+
+    const float low_airspeed = 3.0;
```
I wonder if we should use the VTOL way-point speed here. The presumption being that the aircraft will be stable in 100% VTOL at that speed. Would be nice to remove that jump up to 1:1 scaling when we move out of POS1.

However if we start using this more universally including VTOL modes this might allow pushing WP speed higher and then we have a odd circular dependency. I guess we could always add a new param in the future.

tridge replied:
users don't tend to test if their vehicle is stable at the max WP speed. I think using 3 works well

lthall replied:
I don't care much but I agree with Pete here. You could set the WP speed default to 3.

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -843,7 +843,7 @@ void QuadPlane::run_esc_calibration(void)
 void QuadPlane::multicopter_attitude_rate_update(float yaw_rate_cds)
 {
-    bool use_multicopter_control = in_vtol_mode() && !tailsitter.in_vtol_transition();
+    bool use_multicopter_control = !use_fw_attitude_controllers();
```
I cant take off the H wing with this change.

tridge replied:
i've tested both QLOITER takeoff and AUTO takeoff in the RF9 HWing model with this branch with no issue. Can you give me a log showing a failure?

IamPete1 replied:
@tridge try in Qstab. Qloiter and AUTO don't use this code path. `use_fw_attitude_controllers()` has a `!tailsitter.enabled()` in it so always returns true, hence we never use multi rotor control.

We could puzzle out the exact combination to make it work, but I don't think it should be in this pr. Tailsitters also want to run both controllers sometimes, see the motors only Qassist option.

IamPete1 reviewed (commented):
The changes to `multicopter_attitude_rate_update` and `use_fw_attitude_controllers` break tailsitters.

IamPete1 reviewed (changes requested):
See: https://github.com/ArduPilot/ardupilot/pull/21919#pullrequestreview-1143228300 (I ment to request changes)

Gain scaling stuff looks good.

lthall commented on libraries/AC_AttitudeControl/AC_AttitudeControl.h:
```diff
@@ -376,6 +376,9 @@ class AC_AttitudeControl {
     void set_yaw_rate_tc(float input_tc) { _rate_y_tc = input_tc; }
 
+    // setup a one loop angle P scale
+    void set_angle_P_scale(const Vector3f &angle_P_scale) { _angle_P_scale = angle_P_scale; }
```
I am considering setting this as a product rather than an equal. This would let multiple users access this for different purposes. For example:
Ground resonance suppression
Throttle slew boost

Most of the time everything is just not changing this and would just set it to 1. But this would mean the last user would reset it for the earlier users.

tridge replied:
added both APIs

lthall reviewed (commented):
I have spent some time going over the math behind this and all the ways we want to use this feature and how they might interact and my only feedback is I would like to change the set_angle_P_scale to be a product:

void set_angle_P_scale(const Vector3f &angle_P_scale) { _angle_P_scale *= angle_P_scale; }

Other than this I am really happy with the attitude controller changes here.

tridge commented:
> I would like to change the set_angle_P_scale to be a product

that is easy to do, but I would be concerned we could end up with a code path that does the set twice (or more than twice) and then runs the controller. We'd then apply the square (or cube) of the scale factor.
Can we leave it as an assignment until we have a use case in place where we have two of these scalings running at the same time? When we have that use case we can try to add appropriate protections.
I'm also reluctant to make it a multiply now as I've flown this successfully and wanted it in 4.3.1 to fix a real issue with back transitions. I'd need to organise another test flight if change it to a multiply

tridge commented:
@lthall as discussed I've added a mult version for future use, with comments explaining the difference
I've also added ATSC logging for when it is active

tridge commented:
> * Change in gain is quite "steppy" because of the difference in update rate between the attitude loop and the airspeed loop.

yes, it is steppy, but doesn't seem to have a noticible impact on the actual demanded rates, presumably due to the target filtering
I think this steppiness is fine. we could add another layer of filtering, but that adds lag and complexity

lthall reviewed (approved):
Looks good to me.

IamPete1 reviewed (approved):
Looks good, I like the logging. I think the only input method that does not hit `update_ang_vel_target_from_att_error` is `input_rate_bf_roll_pitch_yaw_2` so in that case we would be logging old values, it might be worth adding:
```
    _angle_P_scale_used = _angle_P_scale;
    _angle_P_scale = Vector3f{1,1,1};
```
To that method too. So the log does not continue with out of date values. (or move the reset to the rate controller run function ;) )

tridge commented:
> so in that case we would be logging old values

I think 

> I think the only input method that does not hit `update_ang_vel_target_from_att_error` is `input_rate_bf_roll_pitch_yaw_2` so in that case we would be logging old values

in that case we're not using the angle P scaling, so we should have reset from the last time we did use them. I agree it's all a bit touchy, but I think it is OK for now
I suspect this code will evolve quite a lot as we use it from more places
